src/Symfony/Component/DomCrawler/Test/Constraint/CrawlerSelectorTextSame.php

```diff
@@ -45,7 +45,13 @@ protected function matches($crawler): bool
             return false;
         }
 
-        return $this->expectedText === trim($crawler->text(null, false));
+        foreach ($crawler as $node) {
+            if ($this->expectedText === trim($node->text(null, false))) {
+                return true;
```

Comment:
for "Contains" I get it's fine to have an "OR" match, but for "Same", shouldn't we expect one single node?

Reply:
I would not expect that it matches the first node only. If I would need that as an requirement, I would add this as a separate assert (with a count probably). I also would expect that assertSelectorTextSame and assertSelectorTextContains behave the same, beside of same/contains logic.

Reply:
I meant "the only one", not "the first".
i.e. when there are two matching nodes, it would fail.
Otherwise, what does "Same" mean?

Reply:
I always understood that _Same_ checks for the whole node value, _Contains_ for a part of it.
Example `<p>Hello world</p>`
**Contains** with selector "p", text "world" would succeed, **Same** would fail.

Reply:
I agree with @arneee .

Same/contains are imho just selectors, they don't indicate something about the number of times it should match. So in that sense, they would act the same as Xpath selectors:

* Same: `//h1[text()="Hello"]`: http://xpather.com/vaNXrvZF
* Contains: `//h1[contains(text(), "Hello")]`: http://xpather.com/bNEbPYag

Reply:
This is clear departure from the current behavior. This means some tests will pass while the might not before.

Reply:
That it is correct, on the other hand it would behave [like the docs say](https://symfony.com/doc/current/testing.html): 

`$this->assertSelectorTextContains('html h1.title', 'Hello World');`
> Using native PHPUnit methods, **the same assertion** would look like this:
```
$this->assertGreaterThan(
    0,
    $crawler->filter('html h1.title:contains("Hello World")')->count()
);
```
(which checks that at least one of the nodes contains the given text)

Reply:
My point is about `assertSelectorTextSame`. The doc doesn't tell about it :)

Reply:
I don't really understand that. Currently, only the first matching element is checked. So if there were more than 1 element that contained the exact same text, it would pass (as only the first one would be checked).

If we now change this assertion to check all matching elements and only allow one to contain this exact text, it would fail.

Reply:
Maybe it should stay as is then... I don't see why this should now pass if the 2nd element matches exactly, while it doesn't right now...

Reply:
When writing tests what would really confuse me... contains / same should imho behave the same regarding number of matches, the only difference should be how the actual node value is checked. Otherwise it is definitely not self-explanatory.

Reply:
Here, we're not designing a new API. We're trying to fix an existing one.
Changing a behavior should not be taken lightly.

Reply:
Fixing the doc might be the better option. Please consider twice.

Reply:
What about adding an `$exactTimes` parameter that defaults to 1?

The issue now is that if we do `assertSelectorTextContains('h3', 'Hello World!')` and we have e.g. a `<h3>Homepage</h3>` in the header (before the `<h3>Hello World!</h3>` in the body), we can't use any of these nice selectors. This is both not nice DX and unexpected behavior imho.

However, I see your point that any change would deviate from the previous behavior and tests that would fail before, can pass now.

Reply:
Adding an argument is a new feature :)